README.md
# osixia/mariadb

[![Docker Pulls](https://img.shields.io/docker/pulls/osixia/mariadb.svg)][hub]
[![Docker Stars](https://img.shields.io/docker/stars/osixia/mariadb.svg)][hub]
[![](https://images.microbadger.com/badges/image/osixia/mariadb.svg)](http://microbadger.com/images/osixia/mariadb "Get your own image badge on microbadger.com")

[hub]: https://hub.docker.com/r/osixia/mariadb/

Latest release: 10.4.11 - MariaDB 10.4.11 - [Changelog](CHANGELOG.md) | [Docker Hub](https://hub.docker.com/r/osixia/mariadb/) 

**A docker image to run MariaDB with perdiodic backups.**
> [MariaDB](https://mariadb.org/)

> [MariaDB Backup](https://mariadb.com/kb/en/library/mariadb-backup/)

- [osixia/mariadb](#osixiamariadb)
	- [Quick start](#quick-start)
	- [Beginner Guide](#beginner-guide)
		- [Create new database](#create-new-database)
			- [Full example](#full-example)
			- [Data persistence](#data-persistence)
		- [Use an existing MariaDB database](#use-an-existing-mariadb-database)
		- [Administrate your mariadb server](#administrate-your-mariadb-server)
		- [Use a custom my.cnf](#use-a-custom-mycnf)
		- [SSL](#ssl)
			- [Use autogenerated certificate](#use-autogenerated-certificate)
			- [Use your own certificate](#use-your-own-certificate)
			- [Disable SSL](#disable-ssl)
		- [Fix docker mounted file problems](#fix-docker-mounted-file-problems)
		- [Debug](#debug)
	- [Environment Variables](#environment-variables)
		- [Set your own environment variables](#set-your-own-environment-variables)
			- [Use command line argument](#use-command-line-argument)
			- [Link environment file](#link-environment-file)
			- [Make your own image or extend this image](#make-your-own-image-or-extend-this-image)
	- [Advanced User Guide](#advanced-user-guide)
		- [Extend osixia/mariadb:10.4.11 image](#extend-osixiamariadb10411-image)
		- [Make your own MariaDB image](#make-your-own-mariadb-image)
		- [Tests](#tests)
		- [Kubernetes](#kubernetes)
		- [Under the hood: osixia/light-baseimage](#under-the-hood-osixialight-baseimage)
	- [Security](#security)
	- [Changelog](#changelog)

## Quick start
Run MariaDB docker image:

	docker run --name my-mariadb-container --detach osixia/mariadb:10.4.11

This start a new container with a MariaDB server running inside.

First we run a terminal on this container:

	docker exec -it my-mariadb-container bash

You should now be in the container terminal,
to properly use this terminal we need to fix the TERM environment variable :

	export TERM=xterm

We can now connect to the MariaDB server using mysql command line tool :

	mysql -u admin -padmin


## Beginner Guide

### Create new database
This is the default behavior when you run the image.

It will create an empty database, with a root and a debian-sys-maint user (required by MariaDB to run properly on debian).
The default root username (admin) and password (admin) can be changed at the docker command line, for example :

	docker run --env MARIADB_ROOT_USER=JaxTeller --env MARIADB_ROOT_PASSWORD=SonsOfAnarchy --detach osixia/mariadb

For security reasons, by default the root user can only login to MariaDB from local networks.
This can also be changed at the docker command line.

For example if you want to allow MariaDB root login from docker default network and localhost :

	docker run --env MARIADB_ROOT_ALLOWED_NETWORKS="#PYTHON2BASH:['172.17.%.%', 'localhost', '127.0.0.1', '::1']" \
	--detach osixia/mariadb


#### Full example
This example will run a docker MariaDB container and execute an sql query from docker host:

	CONTAINER_ID=$(docker run --env MARIADB_ROOT_USER=JaxTeller \
		--env MARIADB_ROOT_PASSWORD=SonsOfAnarchy \
		--env MARIADB_ROOT_ALLOWED_NETWORKS="#PYTHON2BASH:['172.17.%.%', 'localhost', '127.0.0.1', '::1']" \
		--detach osixia/mariadb)

	CONTAINER_IP=$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" $CONTAINER_ID)

	mysql -u JaxTeller -pSonsOfAnarchy -h $CONTAINER_IP -e "select user,host from mysql.user"


#### Data persistence

The directory `/var/lib/mysql` (witch contains all MariaDB database files) has been declared as a volume, so your database files are saved outside the container in a data volume.

For more information about docker data volume, please refer to :

> [https://docs.docker.com/userguide/dockervolumes/](https://docs.docker.com/userguide/dockervolumes/)

### Use an existing MariaDB database

This can be achieved by mounting a host directory as volume.
Assuming you have a MariaDB database on your docker host in the directory `/data/mariadb/CoolDb`
simply mount this directory as a volume to `/var/lib/mysql` :

	docker run --volume /data/mariadb/CoolDb:/var/lib/mysql \
	--env MARIADB_ROOT_USER=MyCoolDbRootUser \
	--env MARIADB_ROOT_PASSWORD=MyCoolDbRootPassword \
	--detach osixia/mariadb

You can also use data volume containers. Please refer to :
> [https://docs.docker.com/userguide/dockervolumes/](https://docs.docker.com/userguide/dockervolumes/)

### Administrate your mariadb server
If you are looking for a simple solution to administrate your MariaDB server you can take a look at our phpMyAdmin docker image:
> [osixia/phpmyadmin](https://github.com/osixia/docker-phpMyAdmin)


### Use a custom my.cnf

Add your custom **my.cnf** in the directory **image/service/mariadb/assets** and rebuild the image ([see manual build](#manual-build)).

Or you can set your custom config at run time, by mounting your **my.cnf** file to **/container/service/mariadb/assets/my.cnf**

	docker run --volume /path/to/my.cnf:/container/service/mariadb/assets/my.cnf --detach osixia/mariadb

### SSL

#### Use autogenerated certificate
By default SSL is enable, a certificate is created with the container hostname (it can be set by docker run --hostname option eg: db.my-company.com).

	docker run --hostname db.my-company.com --detach osixia/mariadb:10.4.11

#### Use your own certificate

You can set your custom certificate at run time, by mounting a directory containing those files to **/container/service/mariadb/assets/certs** and adjust their name with the following environment variables:

	docker run --volume /path/to/certificates:/container/service/mariadb/assets/certs \
	--env MARIADB_SSL_CRT_FILENAME=my-cert.crt \
	--env MARIADB_SSL_KEY_FILENAME=my-cert.key \
	--env MARIADB_SSL_CA_CRT_FILENAME=the-ca.crt \
	--detach osixia/mariadb:10.4.11

Other solutions are available please refer to the [Advanced User Guide](#advanced-user-guide)

#### Disable SSL
Add --env MARIADB_SSL=false to the run command :

    docker run --env MARIADB_SSL=false --detach osixia/mariadb:10.4.11

### Fix docker mounted file problems

You may have some problems with mounted files on some systems. The startup script try to make some file adjustment and fix files owner and permissions, this can result in multiple errors. See [Docker documentation](https://docs.docker.com/v1.4/userguide/dockervolumes/#mount-a-host-file-as-a-data-volume).

To fix that run the container with `--copy-service` argument :

		docker run [your options] osixia/mariadb:10.4.11 --copy-service


### Debug

The container default log level is **info**.
Available levels are: `none`, `error`, `warning`, `info`, `debug` and `trace`.

Example command to run the container in `debug` mode:

	docker run --detach osixia/mariadb:10.4.11 --loglevel debug

See all command line options:

	docker run osixia/mariadb:10.4.11 --help


## Environment Variables

Environment variables defaults are set in **image/environment/default.yaml**

See how to [set your own environment variables](#set-your-own-environment-variables)

Used when the container is started without an existing database:
- **MARIADB_ROOT_USER**: The database root username. Defaults to `admin`
- **MARIADB_ROOT_PASSWORD**: The database root password. Defaults to `admin`
- **MARIADB_ROOT_ALLOWED_NETWORKS**: root login will only be allowed from those networks. Defaults to:
	```yaml
	- localhost
  - 127.0.0.1
  - ::1
	```
	If you want to set this variable at docker run command add the tag `#PYTHON2BASH:` and convert the yaml in python:

		docker run --env MARIADB_ROOT_ALLOWED_NETWORKS="#PYTHON2BASH:['localhost','127.0.0.1','::1']" --detach osixia/mariadb:10.4.11

	To convert yaml to python online: http://yaml-online-parser.appspot.com/

- **MARIADB_DATABASES**: databases to be created on image startup. If user/password was supplied (see below) then  user will be granted superuser access (corresponding to GRANT ALL) to this database. You can also specify a user/password per database (see product database example) Defaults to ` ` (none).

	this variable allows multiples values, for example :
	```yaml
	  - products:
	    - user: password
	    - user2: passw0rd
	  - posts
	  - tomatoes
	```
	If you want to set this variable at docker run command add the tag `#PYTHON2BASH:` and convert the yaml in python:

		docker run --env MARIADB_DATABASES="#PYTHON2BASH:[{'products': [{'user': 'password'}, {'user2': 'passw0rd'}]},'posts','tomatoes']" --detach osixia/mariadb:10.4.11

	To convert yaml to python online: http://yaml-online-parser.appspot.com/

- **MARIADB_USERS**: create a new users with corresponding passwords. These users will be granted superuser permissions (see above) for the databases specified by the MARIADB_DATABASES variable. Defaults to ` ` (none).

	this variable allows multiples values set has user: password for example :
	```yaml
  - boby: mcD0nald
  - billy: th3k1ng
  - tomatoes: ketchup
	```
	If you want to set this variable at docker run command add the tag `#PYTHON2BASH:` and convert the yaml in python:

		docker run --env MARIADB_DATABASES="#PYTHON2BASH:[{'boby': 'mcD0nald'},{'billy': 'th3k1ng'},{'tomatoes': 'ketchup'}]" --detach osixia/mariadb:10.4.11

	To convert yaml to python online: http://yaml-online-parser.appspot.com/

Backup :

- **MARIADB_BACKUP_USER**: The database backup user username. Defaults to `backup`
- **MARIADB_BACKUP_PASSWORD**: The database backup user password. Defaults to `backup`

- **MARIADB_BACKUP_CRON_EXP**: Cron expression to schedule data backup. Defaults to `0 4 * * *`. Every days at 4am.

- **MARIADB_BACKUP_TTL**: Backup TTL in days. Defaults to `15`.

SSL :

- **MARIADB_SSL**: Enable ssl. Defaults to `true`
- **MARIADB_SSL_CIPHER_SUITE**: TLS cipher suite. Defaults to `TLSv1.2`
- **MARIADB_SSL_CRT_FILENAME**: MariaDB ssl certificate filename. Defaults to `mariadb.crt`
- **MARIADB_SSL_KEY_FILENAME**: MariaDB ssl certificate private key filename. Defaults to `mariadb.key`
- **MARIADB_SSL_CA_CRT_FILENAME**: MariaDB ssl CA certificate filename. Defaults to `ca.crt`

	More information at : https://mariadb.com/kb/en/mariadb/ssl-system-variables/

Other environment variables:
- **MARIADB_SSL_HELPER_PREFIX**: ssl-helper environment variables prefix. Defaults to `database`, ssl-helper first search config from DATABASE_SSL_HELPER_* variables, before SSL_HELPER_* variables.

### Set your own environment variables

#### Use command line argument
Environment variables can be set by adding the --env argument in the command line, for example:

	docker run --env MARIADB_ROOT_USER="JaxTeller" --env MARIADB_ROOT_PASSWORD="Sons Of Anarchy" --detach osixia/mariadb:10.4.11

#### Link environment file

For example if your environment file is in :  /data/environment/my-env.yaml

	docker run --volume /data/environment/my-env.yaml:/container/environment/01-custom/env.yaml \
	--detach osixia/mariadb:10.4.11

Take care to link your environment file to `/container/environment/XX-somedir` (with XX < 99 so they will be processed before default environment files) and not  directly to `/container/environment` because this directory contains predefined baseimage environment files to fix container environment (INITRD, LANG, LANGUAGE and LC_CTYPE).

#### Make your own image or extend this image

This is the best solution if you have a private registry. Please refer to the [Advanced User Guide](#advanced-user-guide) just below.

## Advanced User Guide

### Extend osixia/mariadb:10.4.11 image

If you need to add your custom TLS certificate, bootstrap config or environment files the easiest way is to extends this image.

Dockerfile example:

    FROM osixia/mariadb:10.4.11
    MAINTAINER Your Name <your@name.com>

    ADD ssl-certs /container/service/mariadb/assets/certs
    ADD my.cnf /container/service/mariadb/assets/my.cnf
    ADD environment /container/environment/01-custom


### Make your own MariaDB image


Clone this project :

	git clone https://github.com/osixia/docker-mariadb
	cd docker-mariadb

Adapt Makefile, set your image NAME and VERSION, for example :

	NAME = osixia/mariadb
	VERSION = 10.2.6

	becomes :
	NAME = billy-the-king/mariadb
	VERSION = 0.1.0

Add your custom certificate, environment files, my.cnf ...

Build your image :

	make build

Run your image :

	docker run --detach billy-the-king/mariadb:0.1.0

### Tests

We use **Bats** (Bash Automated Testing System) to test this image:

> [https://github.com/sstephenson/bats](https://github.com/sstephenson/bats)

Install Bats, and in this project directory run :

	make test

### Kubernetes

Kubernetes is an open source system for managing containerized applications across multiple hosts, providing basic mechanisms for deployment, maintenance, and scaling of applications.

More information:
- http://kubernetes.io
- https://github.com/kubernetes/kubernetes

A kubernetes example is available in **example/kubernetes**

### Under the hood: osixia/light-baseimage

This image is based on osixia/light-baseimage.
More info: https://github.com/osixia/docker-light-baseimage

## Security
If you discover a security vulnerability within this docker image, please send an email to the Osixia! team at security@osixia.net. For minor vulnerabilities feel free to add an issue here on github.

Please include as many details as possible.

## Changelog

Please refer to: [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project follows MariaDB versioning.

## [10.4.11] - 2019-12-30
### Changed
  - Upgrade MariaDB version to 10.4.11

## [10.4.10] - 2019-11-15
### Changed
  - Upgrade MariaDB version to 10.4.10

## [10.4.8] - 2019-09-28
## Added
  - Multiarch support

### Changed
  - Upgrade MariaDB version to 10.4.8
  - Upgrade baseimage to light-baseimage:1.2.0 (debian buster)

## [10.3.15] - 2019-05-29
### Changed
  - Upgrade MariaDB version to 10.3.15

## [10.3.14] - 2019-04-06
### Changed
  - Upgrade MariaDB version to 10.3.14

## [10.3.13-1] - 2019-04-06
### Fixed
  - Cron log with /usr/bin/logger

## [10.3.13] - 2019-03-21
### Changed
  - Upgrade MariaDB version to 10.3.13
  - max_allowed_packet=64M

### Removed
  - local-infile=0

## [10.3.12] - 2019-01-21
### Changed
  - Upgrade MariaDB version to 10.3.12

## [10.3.10] - 2018-11-08
### Changed
  - Upgrade MariaDB version to 10.3.10

## [10.3.8] - 2018-07-29
### Changed
  - Upgrade MariaDB version to 10.3.8
  - Now use mariabackup instead of xtrabackup

## [10.2.14-1] - 2018-04-09
### Fixed
  - User and database init

## [10.2.14] - 2018-03-28
### Changed
  - Upgrade MariaDB version to 10.2.14

## [10.2.13] - 2018-02-14
### Changed
  - Upgrade MariaDB version to 10.2.13

## [10.2.12] - 2018-01-16
### Changed
  - Upgrade MariaDB version to 10.2.12

## [10.2.11] - 2017-11-29
### Changed
  - Upgrade MariaDB version to 10.2.11

## [10.2.10] - 2017-11-17
### Changed
  - Upgrade MariaDB version to 10.2.10

## [10.2.9] - 2017-11-09
### Changed
  - Upgrade MariaDB version to 10.2.9
  - Upgrade baseimage to light-baseimage:1.1.1

## [10.2.8] - 2017-10-06
### Changed
  - Upgrade MariaDB version to 10.2.8
  - Move config to assets/config folder
  - Use UTF8 by default

## [10.2.6] - 2017-07-19
### Added
  - MARIADB_DATABASES and MARIADB_USERS environment variable, to create databases and users.

### Changed
  - Upgrade MariaDB version to 10.2.6
  - Upgrade baseimage to light-baseimage:1.1.0 (debian stretch)

## [10.1.23] - 2017-05-09
### Changed
  - Upgrade MariaDB version to 10.1.23

## [10.1.22] - 2017-03-15
### Changed
  - Upgrade MariaDB version to 10.1.22

### Fixed
  - Restoration of not running database

## [10.1.21] - 2017-02-18
### Changed
  - Upgrade MariaDB version to 10.1.21

## [10.1.20] - 2016-12-16
### Changed
  - Upgrade MariaDB version to 10.1.20

## [10.1.19] - 2016-11-08
### Added
  - Backups are now compressed and chmod 600

### Changed
  - Upgrade xtrabackup version to 2.4
  - Upgrade MariaDB version to 10.1.19
  - Upgrade baseimage to light-baseimage:0.2.6

## [10.1.18] - 2016-10-15
### Changed
  - Upgrade MariaDB version to 10.1.18

## Versions before following the MariaDB versioning

## [0.2.11] - 2016-09-02
### Changed
  - Upgrade baseimage to light-baseimage:0.2.5
  - Upgrade MariaDB version to 10.1.17

## [0.2.10] - 2016-07-26
### Changed
  - Upgrade baseimage to light-baseimage:0.2.4
  - Upgrade MariaDB version to 10.1.16

## [0.2.9] - 2016-02-20
### Changed
  - Upgrade baseimage to light-baseimage:0.2.2
  - Upgrade MariaDB version to 10.1.11

## [0.2.8] - 2015-12-16
### Added
  - Makefile with build no cache

### Changed
  - Upgrade baseimage to light-baseimage:0.2.0
  - Upgrade MariaDB version to 10.1.9

## [0.2.7] - 2015-11-20
### Changed
  - Upgrade baseimage to light-baseimage:0.1.5

## [0.2.6] - 2015-11-19
### Changed
  - Upgrade baseimage to light-baseimage:0.1.4
  - More easy configuration

## [0.2.5] - 2015-10-23
### Added
  - Add ssl support
  - Add xtrabackup

### Changed
  - MariaDB version 10.1.8
  - Use light-baseimage:0.1.2

## [0.2.4] - 2015-03-03
### Changed
  - Allow single MARIADB_ROOT_ALLOWED_NETWORKS simply by -e MARIADB_ROOT_ALLOWED_NETWORKS=host instead of -e MARIADB_ROOT_ALLOWED_NETWORKS=['host']

## [0.2.3] - 2015-03-02
### Added
  - Custom my.cnf

## [0.2.2] - 2015-02-23
### Changed
  - Upgrade baseimage to baseimage:10.2

### Fixed
  - Install bugs

## [0.2.1] - 2015-01-23
### Added
  - Use Bats as testing tools

### Changed
  - Upgrade baseimage to baseimage:10.1
  - Simplify project tree

## [0.2.0] - 2014-10-15
### Added
  - Add changelog

### Changed
  - Upgrade baseimage to baseimage:0.9.0
  - Change docker command from docker.io to docker

[10.4.11]: https://github.com/osixia/docker-mariadb/compare/v10.4.10...v10.4.11
[10.4.10]: https://github.com/osixia/docker-mariadb/compare/v10.4.8...v10.4.10
[10.4.8]: https://github.com/osixia/docker-mariadb/compare/v10.3.15...v10.4.8
[10.3.15]: https://github.com/osixia/docker-mariadb/compare/v10.3.14...v10.3.15
[10.3.14]: https://github.com/osixia/docker-mariadb/compare/v10.3.13-1...v10.3.14
[10.3.13-1]: https://github.com/osixia/docker-mariadb/compare/v10.3.13...v10.3.13-1
[10.3.13]: https://github.com/osixia/docker-mariadb/compare/v10.3.12...v10.3.13
[10.3.12]: https://github.com/osixia/docker-mariadb/compare/v10.3.10...v10.3.12
[10.3.10]: https://github.com/osixia/docker-mariadb/compare/v10.3.8...v10.3.10
[10.3.8]: https://github.com/osixia/docker-mariadb/compare/v10.2.14-1...v10.3.8
[10.2.14-1]: https://github.com/osixia/docker-mariadb/compare/v10.2.14...v10.2.14-1
[10.2.14]: https://github.com/osixia/docker-mariadb/compare/v10.2.13...v10.2.14
[10.2.13]: https://github.com/osixia/docker-mariadb/compare/v10.2.12...v10.2.13
[10.2.12]: https://github.com/osixia/docker-mariadb/compare/v10.2.11...v10.2.12
[10.2.11]: https://github.com/osixia/docker-mariadb/compare/v10.2.10...v10.2.11
[10.2.10]: https://github.com/osixia/docker-mariadb/compare/v10.2.9...v10.2.10
[10.2.9]: https://github.com/osixia/docker-mariadb/compare/v10.2.8...v10.2.9
[10.2.8]: https://github.com/osixia/docker-mariadb/compare/v10.2.6...v10.2.8
[10.2.6]: https://github.com/osixia/docker-mariadb/compare/v10.1.23...v10.2.6
[10.1.23]: https://github.com/osixia/docker-mariadb/compare/v10.1.22...v10.1.23

image/service/mariadb/startup.sh
#!/bin/bash -e

# set -x (bash debug) if log level is trace
# https://github.com/osixia/docker-light-baseimage/blob/stable/image/tool/log-helper
log-helper level eq trace && set -x

# fix permissions and ownership of /var/lib/mysql
chown -R mysql:mysql /var/lib/mysql
chmod 700 /var/lib/mysql
chown -R mysql:mysql "${CONTAINER_SERVICE_DIR}/mariadb"

# config sql queries
TEMP_FILE='/tmp/mysql-start.sql'

ln -sf "${CONTAINER_SERVICE_DIR}"/mariadb/assets/config/conf.d/* /etc/mysql/conf.d/

FIRST_START_DONE="${CONTAINER_STATE_DIR}/docker-mariadb-first-start-done"
# container first start
if [ ! -e "${FIRST_START_DONE}" ]; then

  #
  # SSL config
  #
  if [ "${MARIADB_SSL,,}" == "true" ]; then
    log-helper info "SSL config..."

    # generate a certificate and key with ssl-helper if LDAP_CRT and LDAP_KEY files don't exists
    # https://github.com/osixia/docker-light-baseimage/blob/stable/image/service-available/:ssl-tools/assets/tool/ssl-helper
    ssl-helper "${MARIADB_SSL_HELPER_PREFIX}" "${CONTAINER_SERVICE_DIR}/mariadb/assets/certs/${MARIADB_SSL_CRT_FILENAME}" "${CONTAINER_SERVICE_DIR}/mariadb/assets/certs/${MARIADB_SSL_KEY_FILENAME}" "${CONTAINER_SERVICE_DIR}/mariadb/assets/certs/${MARIADB_SSL_CA_CRT_FILENAME}"
    chown -R mysql:mysql "${CONTAINER_SERVICE_DIR}/mariadb"
  fi

  #
  # We have a custom config file
  #
  if [ -e "${CONTAINER_SERVICE_DIR}/mariadb/assets/config/my.cnf" ]; then

    log-helper info "Use config file: ${CONTAINER_SERVICE_DIR}/mariadb/assets/config/my.cnf ..."
    rm /etc/mysql/my.cnf
    ln -sf "${CONTAINER_SERVICE_DIR}/mariadb/assets/config/my.cnf" /etc/mysql/my.cnf

  #
  # Use mariadb default config file
  #
  else
    log-helper info "Use default config file: /etc/mysql/my.cnf ..."

    # Allow remote connection
    sed -Ei --follow-symlinks 's/^(bind-address|log)/#&/' /etc/mysql/my.cnf
    # Disable local files loading, don't reverse lookup hostnames, they are usually another container
    sed -i --follow-symlinks '/\[mysqld\]/a\skip-host-cache\nskip-name-resolve' /etc/mysql/my.cnf
  fi

  #
  # there is no database
  #
  if [ -z "$(ls -A -I lost+found /var/lib/mysql)" ]; then

    log-helper info "Init new database..."

    # initializes the MySQL data directory and creates the system tables that it contains
    mysql_install_db --user=mysql --datadir=/var/lib/mysql --rpm

    # start MariaDB
    log-helper info "Start MariaDB..."
    /usr/bin/mysqld_safe --skip-networking --socket=/var/run/mysqld/mysqld.sock &

    mysql=( mysql --protocol=socket -uroot -hlocalhost --socket="/var/run/mysqld/mysqld.sock" )

    for i in {30..0}; do
			if echo 'SELECT 1' | "${mysql}" &> /dev/null; then
				break
			fi
			log-helper info "MySQL init process in progress..."
			sleep 1
		done

    if [ "$i" = 0 ]; then
			log-helper error "MySQL init process failed."
			exit 1
		fi

    # drop all user and test database
    cat > "$TEMP_FILE" <<-EOSQL
        DELETE FROM mysql.user WHERE user NOT IN ('mysql.sys', 'mysqlxsys', 'root') OR host NOT IN ('localhost') ;
        DROP DATABASE IF EXISTS test ;
EOSQL

    # add root user on specified networks
    for network in $(complex-bash-env iterate MARIADB_ROOT_ALLOWED_NETWORKS)
    do
      echo "GRANT ALL PRIVILEGES ON *.* TO '$MARIADB_ROOT_USER'@'${!network}' IDENTIFIED BY '$MARIADB_ROOT_PASSWORD' WITH GRANT OPTION ;" >> "$TEMP_FILE"
    done


    function addUsers() {
      local users=$1
      local databases=$2

      for user in $(complex-bash-env iterate "${users}")
      do
        if [ -n "${!user}" ]; then
          if [ "$(complex-bash-env isRow "${!user}")" = true ]; then
            u=$(complex-bash-env getRowKey "${!user}")
            p=$(complex-bash-env getRowValue "${!user}")

            echo "CREATE USER '$u'@'%' IDENTIFIED BY '$p' ;" >> "$TEMP_FILE"

            for database in $(complex-bash-env iterate "${databases}")
            do
              if [ "$(complex-bash-env isRow "${!database}")" = true ]; then
                database=$(complex-bash-env getRowKeyVarName "${!database}")
              fi

              if [ -n "${!database}" ]; then
                echo "GRANT ALL ON \`${!database}\`.* TO '$u'@'%' ;"  >> "$TEMP_FILE"
              else
                echo "GRANT ALL ON *.* TO '$u'@'%' ;"  >> "$TEMP_FILE"
              fi

            done

          else
            echo "Error please set a password for user: ${!user}"
            exit 1
          fi
        fi
      done
    }

    # add databases
    for database in $(complex-bash-env iterate MARIADB_DATABASES)
    do
      users=""
      # this datase has users
      if [ "$(complex-bash-env isRow "${!database}")" = true ]; then
        users=$(complex-bash-env getRowValueVarName "${!database}")
        database=$(complex-bash-env getRowKeyVarName "${!database}")
      fi

      if [ -n "${!database}" ]; then
        echo "CREATE DATABASE IF NOT EXISTS \`${!database}\` ;" >> "$TEMP_FILE"
      fi

      if [ -n "${users}" ]; then
        # add database specific users
        addUsers "${users}" "${database}"
      fi
    done

    # add global users
    addUsers MARIADB_USERS MARIADB_DATABASES

    # add backup user
    {
      echo "CREATE USER '${MARIADB_BACKUP_USER}'@'localhost' IDENTIFIED BY '${MARIADB_BACKUP_PASSWORD}';"
       echo "GRANT PROCESS, RELOAD, LOCK TABLES, REPLICATION CLIENT ON *.* TO '${MARIADB_BACKUP_USER}'@'localhost';"
    } >> "${TEMP_FILE}"

    # flush privileges
    echo "FLUSH PRIVILEGES ;" >> "${TEMP_FILE}"

    log-helper info "Add MariaDB config..."
    log-helper debug < ${TEMP_FILE}

    # execute config queries
    ${mysql} < ${TEMP_FILE}

    rm ${TEMP_FILE}

    log-helper info "Stop MariaDB..."
    MARIADB_PID=$(cat /var/run/mysqld/mysqld.pid)
    kill -15 "${MARIADB_PID}"
    while [ -e "/proc/${MARIADB_PID}" ]; do sleep 0.1; done # wait until mariadb is terminated

  fi

  touch "${FIRST_START_DONE}"
fi

exit 0
